File: 백준/Silver/10655. 마라톤 1/마라톤 1.py
"""
전체 거리 구하고

점 하나 정하면

그 앞뒤 거리 빼고 그 둘 사이의 거리를 더해서 answer에 넣어두기
"""
import sys

input = lambda: sys.stdin.readline().strip()

N = int(input())

x = []
y = []

for i in range(N):
    a, b = map(int, input().split())
    x.append(a)
    y.append(b)

total = 0
for i in range(N-1):
    total += abs(x[i] - x[i+1]) + abs(y[i] - y[i+1])

answer = total
for i in range(1, N-1):
    result = total
    result -= abs(x[i - 1] - x[i]) + abs(y[i - 1] - y[i])
    result -= abs(x[i] - x[i + 1]) + abs(y[i] - y[i + 1])
    result += abs(x[i - 1] - x[i + 1]) + abs(y[i - 1] - y[i + 1])
    answer = min(answer, result)

print(answer)

File: 백준/Silver/10655. 마라톤 1/마라톤 1.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		int N = Integer.parseInt(br.readLine());
		
		int[] x = new int[N];
		int[] y = new int[N];
		
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			x[i] = Integer.parseInt(st.nextToken());
			y[i] = Integer.parseInt(st.nextToken());
		}
		
		int total = 0;
		for (int i = 0; i < N-1; i++) {
			total += Math.abs(x[i] - x[i+1]) + Math.abs(y[i] - y[i+1]);
		}
		
		int result = 0;
		int answer = total;
		for (int i = 1; i < N-1; i++) {
			result = total;
			result -= Math.abs(x[i-1] - x[i]) + Math.abs(y[i-1] - y[i]);
			result -= Math.abs(x[i] - x[i+1]) + Math.abs(y[i] - y[i+1]);
			result += Math.abs(x[i-1] - x[i+1]) + Math.abs(y[i-1] - y[i+1]);
			answer = Math.min(answer, result);
		}
		
		System.out.println(answer);
		
	} // end of main
} // end of class